src/topp/SageAdapter.cpp

```diff
@@ -90,6 +96,598 @@ class TOPPSageAdapter :
+                charge_states[deltamass].push_back(charge);
+            }
+        }
+        else
+        {
+            hist_it->second += 1.0;
+
+            auto& charges = charge_states[deltamass];
+            if (std::find(charges.begin(), charges.end(), charge) == charges.end())
+            {
+                num_charges_at_mass[deltamass] += 1.0;
+                charges.push_back(charge);
+            }
+        }
+    }
+}
+
+
+ /*    for (auto& id : pips)
+    {
+      auto& hits = id.getHits();
+
+      for (auto& h : hits)
+      { 
+        //Build histogram of deltamasses with charge states 
+         double expval  =  std::stod(h.getMetaValue("SAGE:ExpMass")); 
+         double calcval = std::stod(h.getMetaValue("SAGE:CalcMass"));
+         double deltamass = expval - calcval; 
+         delta_masses.push_back(deltamass); 
+
+        if (hist.find(deltamass) == hist.end())
+        {
+          bool bucketcheck = true; 
+          bool chargecheck = false; 
+
+          for (map<double, double>::const_iterator mit = hist.begin(); mit != hist.end(); ++mit)
+          {
+            //Check with error tolerance in buckets if already present in histogram 
+            if(deltamass <  mit->first+0.0005 && deltamass >  mit->first-0.0005 && bucketcheck && ((0.05 <  deltamass) || (-0.05 >  deltamass)))
+            {
+              hist[mit->first] += 1.0; 
+              bucketcheck = false; 
+
+              for (auto& x : charge_states[mit->first])
+              {
+                if (x == h.getCharge())
+                { 
+                  chargecheck = true; 
+                }
+              }
+              if (!chargecheck)
+              {
+                num_charges_at_mass[mit->first] += 1.0; 
+                charge_states[mit->first].push_back(h.getCharge()); 
+              }
+            }
+          }
+          if (bucketcheck && ((0.05 <  deltamass) || (-0.05 >  deltamass)))
+          {
+             hist[deltamass] += 1.0;
+             num_charges_at_mass[deltamass] += 1.0; 
+             charge_states[deltamass].push_back(h.getCharge()); 
+          }
+        }
+
+        else
+        {
+          hist[deltamass] += 1.0;
+          bool chargecheck = false; 
+          for (auto& x : charge_states[deltamass])
+          {
+            if(x == h.getCharge())
+            {
+              chargecheck = true; 
+            }
+          }
+          if(!chargecheck)
+          {
+            num_charges_at_mass[deltamass] += 1.0; 
+            charge_states[deltamass].push_back(h.getCharge()); 
+          } 
+        }
+      }
+    } */
+      
+
+  
+
+    pair< vector<double>, pair<mapRatetoMass, map<double, double>>> results; 
+    results.second.first = hist; 
+    results.second.second = num_charges_at_mass; 
+    results.first = sorted_keys; 
+    bool with_smoothing = smoothing; 
+
+    if (with_smoothing)
+    {
+      std::map<double, double> smoothed_hist =  smoothMassSpectrum(hist, 0.00001); 
```

Comment:
note: you seem to smooth the histogram not a mass spectrum. mayne needs a better name

Reply:
yeah works. 
I would probably have used: smoothDeltaMassDistribution